cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -293,11 +334,32 @@ std::shared_ptr<CudaDevice> CudaMemoryManager::cuda_device() const {
   return checked_pointer_cast<CudaDevice>(device_);
 }
 
+Result<std::shared_ptr<Device::SyncEvent>> CudaMemoryManager::MakeDeviceSyncEvent() {
+  ARROW_ASSIGN_OR_RAISE(auto context, cuda_device()->GetContext());
+  ContextSaver set_temporary((CUcontext)(context.get()->handle()));
+
+  // TODO: event creation flags??
+  CUevent ev;
+  CU_RETURN_NOT_OK("cuEventCreate", cuEventCreate(&ev, CU_EVENT_DEFAULT));
+
+  return std::shared_ptr<Device::SyncEvent>(
+      new CudaDevice::SyncEvent(context, new CUevent(ev), [](void* ev) {
+        auto typed_event = reinterpret_cast<CUevent*>(ev);
+        auto result = cuEventDestroy(*typed_event);
+        if (result != CUDA_SUCCESS) {
+          // should we throw? I think that would automatically terminate
+          // if you throw in a destructor. What should we do with this error?
```

Comment:
Some prior art here from RMM for CUDA Streams: https://github.com/rapidsai/rmm/blob/228330543cba4921e6b71a52dd1fa81c63921420/include/rmm/cuda_stream.hpp#L67

They use an assert macro (https://github.com/rapidsai/rmm/blob/228330543cba4921e6b71a52dd1fa81c63921420/include/rmm/detail/error.hpp#L204-L245) which is effectively a no-op at runtime.

This error would stick around until the next time someone tries to check the result of a CUDA call and would pop up then, so I think we should be okay to either add a debug assertion or just do nothing here.

Reply:
log at warning level, please. This is what we do for other [errors in destructors](https://github.com/bkietz/arrow/blob/ccae4488e1a0abfc383a2b22f012636ebd47ee21/cpp/src/arrow/util/io_util.cc#L771-L775)

Reply:
Thoughts on using `DCHECK_OK` instead? Or should I just do the ARROW_LOG(warning)?

Reply:
`DCHECK_OK` should be fine